osu.Game.Rulesets.Taiko/Mods/TaikoModAlternate.cs

```diff
@@ -0,0 +1,135 @@
+
+                return false;
+            }
+
+            bool blockInput;
+
+            var previous = HitObjects.ElementAtOrDefault(HitObjects.IndexOf(hitObject) - 1);
+            if ((ResetAfterDrumRoll.Value && previous is DrumRoll) || previous is Swell)
+                ResetActionStates();
+
+            if (hitObject?.IsStrong ?? false)
+                blockInput = handleStrongHit(hitObject, action);
+            else if (hitObject is Swell)
+                blockInput = false;
+            else if (hitObject is IHasDuration)
+                blockInput = action == prevPressedAction;
+            else
+                blockInput = shouldBlock(action);
+
+            if (!strongHitSucceeded)
+            {
+                prevPressedAction = action;
+                prevActionTime = Interceptor.Time.Current;
+                prevHitObject = hitObject;
+            }
+
+            strongHitSucceeded = false;
+
+            return blockInput;
+        }
+
+        private bool shouldBlock(TaikoAction action)
+        {
+            var blockInput = false;
+
+            switch (Style.Value)
+            {
+                case Playstyle.KDDK:
+                    if (prevPressedAction == TaikoAction.LeftRim || prevPressedAction == TaikoAction.LeftCentre)
+                        blockInput = action == TaikoAction.LeftRim || action == TaikoAction.LeftCentre;
+
+                    if (prevPressedAction == TaikoAction.RightRim || prevPressedAction == TaikoAction.RightCentre)
+                        blockInput = action == TaikoAction.RightRim || action == TaikoAction.RightCentre;
+                    break;
+
+                case Playstyle.KKDD:
+                    blockInput = action == prevPressedAction;
+                    break;
+            }
+
+            return blockInput;
+        }
+
+        private bool handleStrongHit(TaikoHitObject hitObject, TaikoAction action)
+        {
+            if ((hitObject == prevHitObject) && (Interceptor.Time.Current - prevActionTime <= strong_hit_window))
+            {
+                var actionsToCheck = (prevPressedAction, action);
+
+                if (compareUnordered(actionsToCheck, (TaikoAction.LeftRim, TaikoAction.RightRim)) ||
+                    compareUnordered(actionsToCheck, (TaikoAction.LeftCentre, TaikoAction.RightCentre)))
+                {
+                    ResetActionStates();
+                    strongHitSucceeded = true;
+
+                    return false;
+                }
+
+                return true;
+            }
+            else
+                return shouldBlock(action);
+
+            bool compareUnordered((TaikoAction?, TaikoAction) actual, (TaikoAction, TaikoAction) expected)
+                => (actual.Item1 == expected.Item1 && actual.Item2 == expected.Item2) ||
+                   (actual.Item1 == expected.Item2 && actual.Item2 == expected.Item1);
+        }
+
+        public enum Playstyle
+        {
+            KDDK,
```

Comment:
what do these mean?

Reply:
I don't have proper names for them since I'm not really a taiko player. Though they're taiko jargons for keybind arrangements. KDDK is Rim-Centre-Centre-Rim and KKDD is Rim-Rim-Centre-Centre (or vice versa).

Reply:
FWIW, this is what I have locally and intend to push in here when I'm done:

```csharp
        public enum Playstyle
        {
            /// <summary>
            /// Each hand has a rim and a centre button, so alternating is done by changing hands.
            /// Also known in community vernacular as "kddk".
            /// </summary>
            [Description(@"Alternate hands (""kddk"")")]
            AlternateHands,

            /// <summary>
            /// One hand has both rim buttons and the other both centre buttons.
            /// Alternating is done using fingers only.
            /// Also known in community vernacular as "kkdd".
            /// </summary>
            [Description(@"Alternate fingers (""kkdd"")")]
            AlternateFingers
        }
```

Still a little arcane, but maybe a little better?